gateway/src/apicast/policy/headers.lua

```diff
@@ -0,0 +1,129 @@
+local _M = policy.new('Headers policy')
+
+local new = _M.new
+
+local function new_header_value(current_value, value_to_add)
+  local new_value = current_value or {}
+
+  if type(new_value) == 'string' then
+    new_value = { new_value }
+  end
+
+  insert(new_value, value_to_add)
+  return new_value
+end
+
+local function add_request_header(header, value, req_headers)
+  local new_value = new_header_value(req_headers[header], value)
+  ngx.req.set_header(header, new_value)
+end
+
+local function set_request_header(header, value)
+  ngx.req.set_header(header, value)
+end
+
+local function push_request_header(header, value, req_headers)
+  if req_headers[header] then
+    add_request_header(header, value, req_headers)
+  end
+end
+
+local function add_resp_header(header, value)
+  local new_value = new_header_value(ngx.header[header], value)
+  ngx.header[header] = new_value
+end
+
+local function set_resp_header(header, value)
+  ngx.header[header] = value
+end
+
+local function push_resp_header(header, value)
+  if ngx.header[header] then
+    add_resp_header(header, value)
+  end
+end
+
+local command_functions = {
+  request = {
+    add = add_request_header,
+    set = set_request_header,
+    push = push_request_header
+  },
+  response = {
+    add = add_resp_header,
+    set = set_resp_header,
+    push = push_resp_header
+  }
+}
+
+-- header_type can be 'request' or 'response'.
+local function run_commands(commands, header_type)
+  -- This is here to avoid calling ngx.req.get_headers() in every command
+  -- applied to the request headers.
+  local req_headers = {}
+  if header_type == 'request' then
+    req_headers = ngx.req.get_headers() or {}
+  end
+
+  for _, command in ipairs(commands) do
+    local command_func = command_functions[header_type][command.op]
+    command_func(command.header, command.value, req_headers)
+  end
+end
+
+-- Initialize the config so we do not have to check for nulls in the rest of
+-- the code.
+local function init_config(config)
+  local res = config or {}
+  res.request = res.request or {}
+  res.response = res.response or {}
+  return res
+end
+
+--- Initialize a Headers policy
+-- @tparam[opt] table config
+-- @field[opt] request Table with the operations to apply to the request headers
+-- @field[opt] response Table with the operations to apply to the response headers
+-- Each operation is a table with three elements:
+--   1) op: can be 'add', 'set' or 'push'.
+--   2) header
+--   3) value
+-- The add operation:
+--   1) When the header is not set, creates it with the given value.
+--   2) When the header is set, it creates a new header with the same name and
+--      the given value.
+-- The set operation:
+--   1) When the header is not set, creates it with the given value.
+--   2) When the header is set, replaces its value with the given one.
+--   3) Deletes a header when the value is "".
+-- The push operation:
+--   1) When the header is not set, it does nothing.
```

Comment:
I thought `push` would always create new header. Thats why it is **push** no ? Like the array operation (and assumption that no header is an empty array).